compiler/src/dotty/tools/dotc/typer/ErrorReporting.scala

```diff
@@ -180,9 +176,189 @@ object ErrorReporting {
+    //   - a function call:
+    //     @implicitAmbiguous("msg A=${A}")
+    //     implicit def f[A](): String = ...
+    //     implicitly[String] // found: f[Any]()
+    //
+    //   - an eta-expanded function:
+    //     @implicitAmbiguous("msg A=${A}")
+    //     implicit def f[A](x: Int): String = ...
+    //     implicitly[Int => String] // found: x => f[Any](x)
+
+    val call = tpd.closureBody(alt.tree) // the tree itself if not a closure
+    val (_, targs, _) = tpd.decomposeCall(call)
+    val args = resolveTypes(targs)(using ctx.fresh.setTyperState(alt.tstate))
+    userDefinedErrorString(raw, params, args)
+  }
+
+  /** @param rawMsg           Message template with variables, e.g. "Variable A is ${A}"
+   *  @param sym              Symbol of the annotated type or of the method whose parameter was annotated
+   *  @param substituteType   Function substituting specific types for abstract types associated with variables, e.g A -> Int
+   */
+  private def formatAnnotationMessage(rawMsg: String, sym: Symbol, substituteType: Type => Type): String = {
+    val substitutableTypesSymbols = ErrorReporting.substitutableTypeSymbolsInScope(sym)
+
+    userDefinedErrorString(
+      rawMsg,
+      paramNames = substitutableTypesSymbols.map(_.name.unexpandedName.toString),
+      args = substitutableTypesSymbols.map(_.typeRef).map(substituteType)
+    )
+  }
+
+  /** Extracting the message from a method parameter, e.g. in
+   *
+   *  trait Foo
+   *
+   *  def foo(implicit @annotation.implicitNotFound("Foo is missing") foo: Foo): Any = ???
+   */
+  private def userDefinedImplicitNotFoundParamMessage = paramSymWithMethodCallTree.flatMap { (sym, applTree) =>
+    userDefinedMsg(sym, defn.ImplicitNotFoundAnnot).map { rawMsg =>
+      val methodOwner = applTree.symbol.owner
+      val methodOwnerType: Type = applTree match {
+        case Select(qual, _) => qual.tpe
+        case TypeApply(Select(qual, _), _) => qual.tpe
+        case _ => methodOwner.typeRef
+      }
+      val substituteTypesFromMethodOwner = (_: Type).asSeenFrom(methodOwnerType, methodOwner)
+      val substituteMethodTypeParams: Type => Type = applTree match {
+        case TypeApply(_, targs) =>
+          val methodTypeParams = applTree.symbol.paramSymss.flatten.filter(_.isType)
+          val methodTypeArgs = targs.map(_.tpe)
+          _.subst(methodTypeParams, methodTypeArgs)
+        case _ =>
+          identity
+      }
+
+      val substituteType = substituteTypesFromMethodOwner andThen substituteMethodTypeParams
+      formatAnnotationMessage(rawMsg, sym.owner, substituteType)
+    }
+  }
+
+  /** Extracting the message from a type, e.g. in
+   *
+   *  @annotation.implicitNotFound("Foo is missing")
+   *  trait Foo
+   *
+   *  def foo(implicit foo: Foo): Any = ???
+   */
+  private def userDefinedImplicitNotFoundTypeMessage = userDefinedMsg(pt.typeSymbol, defn.ImplicitNotFoundAnnot).map { rawMsg =>
+    val substituteType = (_: Type).asSeenFrom(pt, pt.typeSymbol)
```

Comment:
The second argument of `asSeenFrom` is supposed to be a class so `pt.classSymbol` is more appropriate (if pt is a class type it's the same as typeSymbol, it'll be different for a type alias for example).